applications/RANSApplication/custom_utilities/rans_variable_utilities.cpp

```diff
@@ -241,6 +176,159 @@ void CopyNodalSolutionStepVariablesList(
+                r_node.UnSetLock();
+            }
+        }
+    });
+
+    rModelPart.GetCommunicator().AssembleCurrentData(rVariable);
+}
+
+void AddAnalysisStep(
+    ModelPart& rModelPart,
+    const std::string& rStepName)
+{
+    auto& r_process_info = rModelPart.GetProcessInfo();
+    if (!r_process_info.Has(ANALYSIS_STEPS))
+    {
+        r_process_info.SetValue(ANALYSIS_STEPS, std::vector<std::string>());
+    }
+    r_process_info[ANALYSIS_STEPS].push_back(rStepName);
+}
+
+bool IsAnalysisStepCompleted(
+    const ModelPart& rModelPart,
+    const std::string& rStepName)
+{
+    const auto& r_process_info = rModelPart.GetProcessInfo();
+    if (r_process_info.Has(ANALYSIS_STEPS))
+    {
+        const std::vector<std::string>& r_steps = r_process_info[ANALYSIS_STEPS];
+        return (std::find(r_steps.begin(), r_steps.end(), rStepName) != r_steps.end());
+    }
+    else
+    {
+        return false;
+    }
+}
+
+void AssignBoundaryFlagsToGeometries(
+    ModelPart& rModelPart)
+{
+    block_for_each(rModelPart.Conditions(), [&](ModelPart::ConditionType& rCondition) {
+        rCondition.SetValue(RANS_IS_INLET, rCondition.Is(INLET));
+        rCondition.SetValue(RANS_IS_OUTLET, rCondition.Is(OUTLET));
+        rCondition.SetValue(RANS_IS_STRUCTURE, rCondition.Is(STRUCTURE));
+    });
+}
+
+void CalculateMagnitudeSquareForNodal3DVariable(
+    ModelPart& rModelPart,
+    const Variable<array_1d<double, 3>>& r3DVariable,
+    const Variable<double>& rOutputVariable)
+{
+    block_for_each(rModelPart.Nodes(), [&](ModelPart::NodeType& rNode) {
+        const double magnitude = norm_2(rNode.FastGetSolutionStepValue(r3DVariable));
+        rNode.FastGetSolutionStepValue(rOutputVariable) = std::pow(magnitude, 2);
+    });
+}
+
+template <>
+KRATOS_API(RANS_APPLICATION)
+double GetVariableValueNorm(const double& rValue)
+{
+    return rValue;
+}
+
+template <>
+KRATOS_API(RANS_APPLICATION)
+double GetVariableValueNorm(const array_1d<double, 3>& rValue)
+{
+    return norm_2(rValue);
+}
+
+template <typename TDataType>
+std::tuple<double, double> CalculateTransientVariableConvergence(
+    const ModelPart& rModelPart,
+    const Variable<TDataType>& rVariable)
+{
+    KRATOS_TRY
+
+    const auto& r_communicator = rModelPart.GetCommunicator();
+    const auto& r_nodes = r_communicator.LocalMesh().Nodes();
+    const int number_of_nodes = r_nodes.size();
+
+    KRATOS_ERROR_IF(rModelPart.GetBufferSize() < 2)
+        << rModelPart.Name() << " buffer size is "
+        << rModelPart.GetBufferSize() << ". Buffer size of 2 or greater is required to calculate transient variable convergence for "
+        << rVariable.Name() << ".\n";
+
+    double dx, solution, number_of_dofs;
+    std::tie(dx, solution, number_of_dofs) =
+        IndexPartition<int>(number_of_nodes)
+            .for_each<CombinedReduction<SumReduction<double>, SumReduction<double>, SumReduction<double>>>(
```

Comment:
I think you can't use the `block_for_each` in here.

Reply:
`const ModelPart` doesn't allow it.

Reply:
You're right (I think we already discussed this in another PR).